File: _inc/laravel/app/Http/Controllers/Bills/BenefitPaymentController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{
    DatabaseConstants,
    MiddlewaresConstants,
    SettingsConstants,
    UsersConstants,
    ViewsConstants,
};
use App\Models\{
    Coupon,
    Invoice,
    InvoiceBankTransfer,
    InvoicePayment,
    Order,
    Plan,
    User,
    UserCoupon,
    Utility
};
use App\Traits\ChecksLogin;
use GuzzleHttp\Client;
use Illuminate\Auth\Access\AuthorizationException;
use Illuminate\Http\{
    Request,
    RedirectResponse,
    JsonResponse
};
use Illuminate\Support\Facades\{
    Auth,
    Crypt,
    DB,
    Route,
    Log,
    View as ViewFacade
};
use Throwable;

use function App\Http\Controllers\Helpers\{defaultUndefinedException, defaultPermissionDenial};
final class BenefitPaymentController extends Controller
{
    use ChecksLogin;

    public function __construct()
    {
        $this->middleware([MiddlewaresConstants::AUTH]);
    }

    public const INI_PAY = 'initiatePayment';
    public function initiatePayment(Request $req): RedirectResponse|JsonResponse|null
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class  = static::class;
        $base   = class_basename($class);
        return $this->measureProfile($action, function () use ($req, $action, $method, $base) {
            Log::info("[{$base}::{$action}] start", [UsersConstants::COL_USER_ID => Auth::id(), 'plan_id' => $req->input('plan_id'), 'coupon_raw' => $req->input('coupon'), 'method' => $method]);
            if (($r = self::_checkLogin()) instanceof RedirectResponse) return $r;
            if ($r = self::authorizePerm($req, 'buy plan')) return $r;
            try {
                $cfgStart = microtime(true);
                $settings = Utility::getAdminPaymentSetting();
                $secret   = $settings['benefit_secret_key'] ?? '';
                $this->logExecutionTime($cfgStart, $action, 'getAdminPaymentSetting');
                $user   = $req->user();
                $decStart = microtime(true);
                $planId = Crypt::decryptString($req->input('plan_id'));
                $this->logExecutionTime($decStart, $action, 'decryptPlanId');
                $planStart = microtime(true);
                $plan   = Plan::findOrFail($planId);
                $this->logExecutionTime($planStart, $action, 'findPlan');
                $amount = $plan->price;
                $couponCode = strtoupper($req->input('coupon', ''));
                Log::info("[{$base}::{$action}] plan lookup", ['plan_id' => $planId, 'price' => $amount]);
                if ($couponCode) {
                    $cpFetchStart = microtime(true);
                    $coupon = Coupon::where('code', $couponCode)->where('is_active', 1)->first();
                    $this->logExecutionTime($cpFetchStart, $action, 'fetchCoupon');
                    if (!$coupon) {
                        Log::warning("[{$base}::{$action}] invalid coupon", ['code' => $couponCode]);
                        return back()->with('error', __('This coupon code is invalid or expired.'));
                    }
                    if ($coupon->limit <= $coupon->used_coupon) {
                        Log::warning("[{$base}::{$action}] expired coupon", ['code' => $couponCode]);
                        return back()->with('error', __('This coupon code has expired.'));
                    }
                    $applyStart = microtime(true);
                    $discount = ($plan->price * ($coupon->discount / 100));
                    $amount -= $discount;
                    $this->logExecutionTime($applyStart, $action, 'applyCoupon');
                    Log::info("[{$base}::{$action}] coupon applied", ['code' => $couponCode, 'discount' => $discount, 'new_amount' => $amount]);
                    if ($amount <= 0) {
                        DB::beginTransaction();
                        try {
                            $freeStart = microtime(true);
                            $user?->assignPlan($plan->id);
                            UserCoupon::create(['user' => $user?->id, 'coupon' => $coupon->id, 'order' => self::txnId()]);
                            $coupon->update(['is_active' => 0]);
                            $this->logExecutionTime($freeStart, $action, 'activateFreePlan');
                            DB::commit();
                            Log::info("[{$base}::{$action}] free plan activated via coupon", [UsersConstants::COL_USER_ID => $user?->id, 'plan_id' => $plan->id]);
                            return redirect()->route(ViewsConstants::PLN . '.index')->with('success', __('Plan successfully activated.'));
                        } catch (Throwable $e) {
                            DB::rollBack();
                            throw $e;
                        }
                    }
                }
                $payload = [
                    'amount' => $amount,
                    'currency' => $settings['currency'] ?? 'BHD',
                    'customer_initiated' => true,
                    'threeDSecure' => true,
                    'save_card' => false,
                    'description' => 'Plan - ' . $plan->name,
                    'metadata' => ['udf1' => 'Metadata 1'],
                    'reference' => ['transaction' => 'txn_01', 'order' => 'ord_01'],
                    'receipt' => ['email' => true, 'sms' => true],
                    'customer' => ['first_name' => $user?->name, 'email' => $user?->email, 'phone' => ['country_code' => 965, 'number' => 51234567]],
                    'source' => ['id' => 'src_bh.benefit'],
                    'redirect' => ['url' => route('benefit.callback', ['plan' => $plan->id, 'amount' => $amount, 'coupon' => $couponCode ?: '0'])],
                ];
                Log::info("[{$base}::{$action}] sending charge request", ['payload' => $payload]);
                $httpStart = microtime(true);
                $response = self::client()->post('https://api.tap.company/v2/charges', ['json' => $payload, 'headers' => ['Authorization' => 'Bearer ' . $secret, 'Accept' => 'application/json']]);
                $this->logExecutionTime($httpStart, $action, 'initiateCharge');
                $data = json_decode($response->getBody()->getContents());
                Log::info("[{$base}::{$action}] charge initiated", ['transaction_url' => $data->transaction->url]);
                return redirect($data->transaction->url);
            } catch (Throwable $e) {
                Log::error("[{$base}::{$action}] failed", ['error' => $e->getMessage()]);
                Log::debug("[{$base}::{$action}] debug context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine(), 'code' => $e->getCode(), 'route' => Route::getCurrentRoute()?->getName()]);
                return self::handleException($req, $e);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'plan_id' => $req->input('plan_id'), 'coupon' => $req->input('coupon')]);
    }

    public const CB = 'callBack';
    public function callBack(Request $req): RedirectResponse|JsonResponse|null
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        $request = $req;
        return $this->measureProfile($action, function () use ($request, $action, $method, $base) {
            Log::info("[{$base}::{$action}] callback start", ['tap_id' => $request->input('tap_id'), 'plan' => $request->input('plan'), 'coupon' => $request->input('coupon'), 'method' => $method]);
            if (($r = self::_checkLogin()) instanceof RedirectResponse) return $r;
            try {
                $cfgStart = microtime(true);
                $settings = Utility::getAdminPaymentSetting();
                $secret = $settings['benefit_secret_key'] ?? '';
                $plan = Plan::findOrFail($request->input('plan'));
                $user = $request->user();
                $couponCode = $request->input('coupon') !== '0' ? $request->input('coupon') : null;
                $this->logExecutionTime($cfgStart, $action, 'prepareCallback');
                $httpStart = microtime(true);
                $resp = self::client()->get('https://api.tap.company/v2/charges/' . $request->input('tap_id'), ['headers' => ['Authorization' => 'Bearer ' . $secret, 'Accept' => 'application/json']]);
                $this->logExecutionTime($httpStart, $action, 'tapChargeGet');
                $parseStart = microtime(true);
                $status = json_decode($resp->getBody()->getContents());
                $this->logExecutionTime($parseStart, $action, 'parseTapStatus');
                Log::info("[{$base}::{$action}] charge status", ['code' => $status->gateway->response->code]);
                if ($status->gateway->response->code !== '00') {
                    Log::warning("[{$base}::{$action}] transaction failed", ['code' => $status->gateway->response->code]);
                    Log::debug("[{$base}::{$action}] failure context", ['tap_id' => $request->input('tap_id'), 'plan_id' => $plan->id, 'user_id' => $user?->id]);
                    return redirect()->route(ViewsConstants::PLN . '.index')->with('error', __('Transaction failed, please try again.'));
                }
                $txnStart = microtime(true);
                DB::beginTransaction();
                try {
                    $orderId = self::txnId();
                    $createStart = microtime(true);
                    Order::create([
                        'order_id' => $orderId,
                        'plan_name' => $plan->name,
                        'plan_id' => $plan->id,
                        'price' => $request->input('amount'),
                        'price_currency' => Utility::getValByName('currency'),
                        'payment_type' => 'Benefit',
                        'payment_status' => 'success',
                        UsersConstants::COL_USER_ID => $user?->id,
                    ]);
                    $this->logExecutionTime($createStart, $action, 'createOrder');
                    $assignStart = microtime(true);
                    $user?->assignPlan($plan->id);
                    $this->logExecutionTime($assignStart, $action, 'assignPlan');
                    Log::info("[{$base}::{$action}] plan assigned", [UsersConstants::COL_USER_ID => $user?->id, 'plan_id' => $plan->id]);
                    if ($couponCode) {
                        $cStart = microtime(true);
                        $coupon = Coupon::where('code', $couponCode)->first();
                        $this->logExecutionTime($cStart, $action, 'fetchCoupon');
                        if ($coupon) {
                            $ucStart = microtime(true);
                            UserCoupon::create(['user' => $user?->id, 'coupon' => $coupon->id, 'order' => $orderId]);
                            $this->logExecutionTime($ucStart, $action, 'createUserCoupon');
                            if ($coupon->limit <= $coupon->used_coupon) {
                                $deactStart = microtime(true);
                                $coupon->update(['is_active' => 0]);
                                $this->logExecutionTime($deactStart, $action, 'deactivateCoupon');
                            }
                            Log::info("[{$base}::{$action}] coupon attached", ['coupon' => $couponCode, 'order' => $orderId]);
                        }
                    }
                    DB::commit();
                    $this->logExecutionTime($txnStart, $action, 'transactionCommit');
                    return redirect()->route(ViewsConstants::PLN . '.index')->with('success', __('Plan activated successfully.'));
                } catch (Throwable $e) {
                    DB::rollBack();
                    $this->logExecutionTime($txnStart, $action, 'transactionRollback');
                    Log::error("[{$base}::{$action}] callback failed", ['error' => $e->getMessage()]);
                    Log::debug("[{$base}::{$action}] callback failure context", ['file' => $e->getFile(), 'line' => $e->getLine(), 'trace' => $e->getTraceAsString(), 'user_id' => $user?->id, 'plan_id' => $plan->id]);
                    throw $e;
                }
            } catch (Throwable $e) {
                Log::error("[{$base}::{$action}] exception", ['message' => $e->getMessage()]);
                Log::debug("[{$base}::{$action}] exception context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine(), 'trace' => $e->getTraceAsString()]);
                return self::handleException($request, $e);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public const INV_PAY_BF = 'invoicePayWithBenefit';
    public function invoicePayWithBenefit(Request $req): RedirectResponse|JsonResponse|null
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        $request = $req;
        return $this->measureProfile($action, function () use ($request, $action, $method, $base) {
            Log::info("[{$base}::{$action}] start", ['invoice_id_enc' => $request->input('invoice_id'), 'amount' => $request->input('amount'), 'method' => $method]);
            if (($r = self::_checkLogin()) instanceof RedirectResponse) return $r;
            try {
                $decStart = microtime(true);
                $invoiceId = Crypt::decryptString($request->input('invoice_id'));
                $this->logExecutionTime($decStart, $action, 'decryptInvoiceId');
                $invStart = microtime(true);
                $invoice = Invoice::findOrFail($invoiceId);
                $this->logExecutionTime($invStart, $action, 'findInvoice');
                $cfgStart = microtime(true);
                $settings = Utility::getCompanyPaymentSetting($invoice[DatabaseConstants::COL_TABLE_CREATOR]);
                $secret = $settings['benefit_secret_key'] ?? '';
                $this->logExecutionTime($cfgStart, $action, 'getCompanyPaymentSetting');
                $amount = (float) $request->input('amount');
                Log::info("[{$base}::{$action}] invoice lookup", ['invoice_id' => $invoiceId, 'due' => $invoice->getDue()]);
                if ($amount <= 0 || $amount > $invoice->getDue()) {
                    Log::warning("[{$base}::{$action}] invalid amount", ['amount' => $amount]);
                    Log::debug("[{$base}::{$action}] invalid amount context", ['invoice_id' => $invoiceId, 'due' => $invoice->getDue(), 'user_id' => $request->user()?->id]);
                    return back()->with('error', __('Invalid amount.'));
                }
                $payer = Auth::check() ? Auth::user() : User::find($invoice[DatabaseConstants::COL_TABLE_CREATOR]);
                $charge = [
                    'amount' => $amount,
                    'currency' => Utility::settingsById($invoice[DatabaseConstants::COL_TABLE_CREATOR])['site_currency'] ?? 'BHD',
                    'customer_initiated' => true,
                    'threeDSecure' => true,
                    'save_card' => false,
                    'description' => $invoice->invoice_id,
                    'metadata' => ['udf1' => 'Metadata 1'],
                    'reference' => ['transaction' => 'txn_01', 'order' => 'ord_01'],
                    'receipt' => ['email' => true, 'sms' => true],
                    'customer' => ['first_name' => $payer->name, 'email' => $payer->email, 'phone' => ['country_code' => 965, 'number' => 51234567]],
                    'source' => ['id' => 'src_bh.benefit'],
                    'redirect' => ['url' => route(ViewsConstants::INV . 'benefit.status', ['invoice' => $request->input('invoice_id'), 'amount' => $amount])],
                ];
                Log::info("[{$base}::{$action}] sending invoice charge", ['payload' => $charge]);
                $httpStart = microtime(true);
                $response = self::client()->post('https://api.tap.company/v2/charges', ['json' => $charge, 'headers' => ['Authorization' => 'Bearer ' . $secret, 'Accept' => 'application/json']]);
                $this->logExecutionTime($httpStart, $action, 'tapChargePost');
                $parseStart = microtime(true);
                $data = json_decode($response->getBody()->getContents());
                $this->logExecutionTime($parseStart, $action, 'parseTapResponse');
                Log::info("[{$base}::{$action}] invoice charge initiated", ['transaction_url' => $data->transaction->url]);
                return redirect($data->transaction->url);
            } catch (Throwable $e) {
                Log::error("[{$base}::{$action}] exception", ['message' => $e->getMessage()]);
                Log::debug("[{$base}::{$action}] exception context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine(), 'trace' => $e->getTraceAsString(), 'invoice_id_enc' => $request->input('invoice_id')]);
                return self::handleException($request, $e);
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base]);
    }

    public const GET_INV_PAY_STT = 'getInvoicePaymentStatus';
    public function getInvoicePaymentStatus(Request $req, string $invoiceEncrypted, string $amount): RedirectResponse
    {
        $action = __FUNCTION__;
        $method = __METHOD__;
        $class = static::class;
        $base = class_basename($class);
        $request = $req;
        return $this->measureProfile($action, function () use ($request, $invoiceEncrypted, $amount, $action, $method, $base) {
            Log::info("[{$base}::{$action}] start", ['invoice_enc' => $invoiceEncrypted, 'tap_id' => $request->input('tap_id'), 'amount' => $amount, 'method' => $method]);
            if (($r = self::_checkLogin()) instanceof RedirectResponse) return $r;
            try {
                $decStart = microtime(true);
                $invoiceId = Crypt::decryptString($invoiceEncrypted);
                $this->logExecutionTime($decStart, $action, 'decryptInvoiceId');
                $invStart = microtime(true);
                $invoice = Invoice::findOrFail($invoiceId);
                $this->logExecutionTime($invStart, $action, 'findInvoice');
                $usrStart = microtime(true);
                $user = User::findOrFail($invoice[DatabaseConstants::COL_TABLE_CREATOR]);
                $this->logExecutionTime($usrStart, $action, 'findUser');
                $cfgStart = microtime(true);
                $secret = Utility::getCompanyPaymentSetting($user?->id)['benefit_secret_key'] ?? '';
                $this->logExecutionTime($cfgStart, $action, 'getCompanyPaymentSetting');
                $httpStart = microtime(true);
                $resp = self::client()->get('https://api.tap.company/v2/charges/' . $request->input('tap_id'), ['headers' => ['Authorization' => 'Bearer ' . $secret, 'Accept' => 'application/json']]);
                $this->logExecutionTime($httpStart, $action, 'tapChargeGet');
                $parseStart = microtime(true);
                $status = json_decode($resp->getBody()->getContents());
                $this->logExecutionTime($parseStart, $action, 'parseTapStatus');
                Log::info("[{$base}::{$action}] payment status", ['code' => $status->gateway->response->code]);
                if ($status->gateway->response->code !== '00') {
                    Log::warning("[{$base}::{$action}] transaction failed", ['invoice_id' => $invoiceId]);
                    Log::debug("[{$base}::{$action}] failure context", ['tap_id' => $request->input('tap_id'), 'user_id' => $user?->id, 'secret_set' => !empty($secret)]);
                    return redirect()->route(ViewsConstants::INV . 'link.copy', $invoiceEncrypted)->with('error', __('Transaction failed!'));
                }
                $txnStart = microtime(true);
                DB::beginTransaction();
                try {
                    $orderId = self::txnId();
                    $ipStart = microtime(true);
                    InvoicePayment::create([
                        'invoice_id' => $invoice->id,
                        'date' => now()->toDateString(),
                        'amount' => $amount,
                        'account_id' => 0,
                        'payment_method' => 0,
                        'order_id' => $orderId,
                        'payment_type' => 'Benefit',
                        'description' => 'Invoice ' . Utility::invoiceNumberFormat(Utility::settingsById($invoice[DatabaseConstants::COL_TABLE_CREATOR]), $invoice->invoice_id),
                    ]);
                    $this->logExecutionTime($ipStart, $action, 'createInvoicePayment');
                    $stStart = microtime(true);
                    $newDue = $invoice->getDue() - (float) $amount;
                    Invoice::changeStatus($invoice->id, $newDue ? 2 : 3);
                    $this->logExecutionTime($stStart, $action, 'updateInvoiceStatus');
                    $balStart = microtime(true);
                    Utility::updateUserBalance('customer', $invoice->customer_id, $amount, 'debit');
                    $this->logExecutionTime($balStart, $action, 'updateUserBalance');
                    $delStart = microtime(true);
                    InvoiceBankTransfer::where('invoice_id', $invoice->id)->where('order_id', $orderId)->delete();
                    $this->logExecutionTime($delStart, $action, 'cleanupTransfer');
                    DB::commit();
                    $this->logExecutionTime($txnStart, $action, 'transactionCommit');
                    Log::info("[{$base}::{$action}] invoice payment recorded", ['invoice_id' => $invoiceId, 'order_id' => $orderId]);
                    return redirect()->route(ViewsConstants::INV . 'link.copy', $invoiceEncrypted)->with('success', __('Invoice paid successfully!'));
                } catch (Throwable $e) {
                    DB::rollBack();
                    $this->logExecutionTime($txnStart, $action, 'transactionRollback');
                    Log::error("[{$base}::{$action}] recording failed", ['error' => $e->getMessage(), 'invoice_id' => $invoiceId]);
                    Log::debug("[{$base}::{$action}] recording failure context", ['file' => $e->getFile(), 'line' => $e->getLine(), 'trace' => $e->getTraceAsString(), 'amount' => $amount, 'user_id' => $user?->id]);
                    throw $e;
                }
            } catch (Throwable $e) {
                Log::error("[{$base}::{$action}] failed", ['err' => $e->getMessage()]);
                Log::debug("[{$base}::{$action}] exception context", ['exception' => get_class($e), 'file' => $e->getFile(), 'line' => $e->getLine(), 'trace' => $e->getTraceAsString(), 'invoice_enc' => $invoiceEncrypted, 'tap_id' => $request->input('tap_id')]);
                return redirect()->route(ViewsConstants::INV . 'link.copy', $invoiceEncrypted)->with('error', $e->getMessage());
            }
        }, ['route' => Route::getCurrentRoute()?->getName(), 'method' => $method, 'class' => $base, 'invoice_enc' => $invoiceEncrypted, 'amount' => $amount]);
    }

    private static function authorizePerm(Request $req, string $perm): RedirectResponse|JsonResponse|null
    {
        $user = $req->user();
        if ($user?->can($perm)) {
            Log::info(__METHOD__ . " permission granted", [
                UsersConstants::COL_USER_ID => $user?->id,
                'perm' => $perm
            ]);
            return null;
        }
        Log::warning(__METHOD__ . " permission denied", [
            UsersConstants::COL_USER_ID => $user?->id,
            'perm' => $perm
        ]);
        return defaultPermissionDenial(
            $req,
            new AuthorizationException($perm),
            __CLASS__ . '::' . __FUNCTION__
        );
    }

    private static function handleException(Request $req, Throwable $e): RedirectResponse|JsonResponse|null
    {
        Log::error(__METHOD__ . " exception", [
            'message' => $e->getMessage(),
        ]);
        Log::channel(SettingsConstants::ERR_TRACE)->debug(__METHOD__ . " exception", [
            'message' => $e->getMessage(),
            'trace'   => $e->getTraceAsString()
        ]);
        return defaultUndefinedException(
            $req,
            $e,
            __CLASS__ . '::' . __FUNCTION__
        );
    }

    private static function client(): Client
    {
        return new Client(['timeout' => 15]);
    }

    private static function txnId(): string
    {
        return strtoupper(str_replace('.', '', uniqid('', true)));
    }
}

// ! ALERT secret_key is stored in plain‑text settings and sent back in exceptions; avoid exposing it and consider moving to env‑encrypted vault.

File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/bills/benefit_payment_controller.py
import logging
import inspect
import json
import requests
from django.shortcuts import redirect
from django.contrib import messages
from django.http import HttpRequest, HttpResponse
from django.urls import reverse

from ....Models.activity.order import Order
from ....Models.bills.coupon import Coupon
from ....Models.bills.invoice import Invoice
from ....Models.bills.invoice_payment import InvoicePayment
from ....Models.bills.user_coupon import UserCoupon
from ....Models.individuals.customer import Customer
from ....Models.individuals.user import User
from ....Models.planning.plan import Plan
from ....Models.utils.utility import Utility
from .._helpers.http import get_redirect_url
from .._helpers.error_handlers import default_undefined_exception
from .._traits.controller import Controller

logger = logging.getLogger(__name__)

class BenefitPaymentController(Controller):

    @classmethod
    def initiate_payment(cls, request: HttpRequest) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            # settings_dict = Utility.settings()
            admin_settings = Utility.get_admin_payment_setting()
            secret_key = admin_settings.get('benefit_secret_key')
            user = request.user
            plan_id = Utility.decrypt(request.POST.get('plan_id'))
            plan = Plan.objects.filter(id=plan_id).first()
            if not plan:
                return redirect('plans.index')
            price = plan.price
            coupon_code = request.POST.get('coupon', '').strip()
            # coupon_id = None
            if coupon_code:
                coupon = Coupon.objects.filter(
                    code=coupon_code.upper(),
                    is_active=True
                ).first()
                if not coupon:
                    messages.error(request, "This coupon code is invalid or has expired.")
                    return redirect(get_redirect_url(request))
                if coupon.used_coupon() >= coupon.limit:
                    messages.error(request, "This coupon code has expired.")
                    return redirect(get_redirect_url(request))
                discount = (plan.price / 100) * coupon.discount
                price -= discount
                # coupon_id = coupon.id
                if price <= 0:
                    user.plan_id = plan.id
                    user.save()
                    result = user.assign_plan(plan.id)
                    if result.get('is_success'):
                        Order.objects.create(
                            order_id=Utility.generate_order_id(),
                            plan_name=plan.name,
                            plan_id=plan.id,
                            price=0,
                            price_currency=admin_settings.get('currency', 'BHD'),
                            payment_type='Benefit',
                            payment_status='success',
                            user_id=user.id
                        )
                        if user.payment_subscription_id:
                            try:
                                user.cancel_subscription(user.id)
                            except Exception as e:
                                logger.exception(f"Failed to cancel subscription: {e}")
                        UserCoupon.objects.create(
                            user=user.id,
                            coupon=coupon.id,
                            order=Utility.generate_order_id()
                        )
                        return redirect('plans.index')
            payload = {
                'amount': price,
                'currency': admin_settings.get('currency', 'BHD'),
                'customer_initiated': True,
                'threeDSecure': True,
                'save_card': False,
                'description': f'Plan - {plan.name}',
                'metadata': {'udf1': 'Metadata 1'},
                'reference': {'transaction': 'txn_01', 'order': 'ord_01'},
                'receipt': {'email': True, 'sms': True},
                'customer': {
                    'first_name': user.name,
                    'middle_name': '',
                    'last_name': '',
                    'email': user.email,
                    'phone': {'country_code': 965, 'number': '51234567'}
                },
                'source': {'id': 'src_bh.benefit'},
                'post': {'url': 'https://webhook.site/fd8b0712-d70a-4280-8d6f-9f14407b3bbd'},
                'redirect': {
                    'url': reverse('BenefitPaymentController.call_back', kwargs={
                        'plan_id': plan.id,
                        'amount': price,
                        'coupon': coupon_code if coupon_code else '0'
                    })
                }
            }
            res = requests.post(
                'https://api.tap.company/v2/charges',
                headers={
                    'Authorization': f'Bearer {secret_key}',
                    'accept': 'application/json',
                    'content-type': 'application/json',
                },
                data=json.dumps(payload)
            )
            return redirect(res.json()['transaction']['url'])
        except Exception as err:
            return default_undefined_exception(request, err=err, ref=REF, logger=logger)

    @classmethod
    def call_back(cls, request: HttpRequest, plan_id: int,
                  amount: float, coupon: str) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            settings_dict = Utility.get_admin_payment_setting()
            secret_key = settings_dict.get('benefit_secret_key')
            user = request.user
            plan = Plan.objects.get(id=plan_id)
            order_id = Utility.generate_order_id()
            coupon_id = None
            if coupon != '0':
                coupon_obj = Coupon.objects.filter(
                    code=coupon.upper(),
                    is_active=True
                ).first()
                coupon_id = coupon_obj.id if coupon_obj else None
            tap_id = request.GET.get('tap_id')
            res = requests.get(
                f'https://api.tap.company/v2/charges/{tap_id}',
                headers={
                    'Authorization': f'Bearer {secret_key}',
                    'accept': 'application/json'
                }
            )
            data = res.json()
            status_code = data['gateway']['response']['code']
            if status_code == '00':
                Order.objects.create(
                    order_id=order_id,
                    name=user.name,
                    plan_name=plan.name,
                    plan_id=plan.id,
                    price=amount,
                    price_currency=settings_dict.get('currency', 'BHD'),
                    payment_type='Benefit',
                    payment_status='success',
                    user_id=user.id
                )
                if coupon_id:
                    user_coupon = UserCoupon(user=user.id, coupon=coupon_id, order=order_id)
                    user_coupon.save()
                    c_obj = Coupon.objects.get(id=coupon_id)
                    if c_obj.used_coupon() >= c_obj.limit:
                        Coupon.objects.filter(id=coupon_id).update(is_active=False)
                result = user.assign_plan(plan.id)
                messages.success(request, 'Plan activated Successfully.') if result.get('is_success') \
                    else messages.error(request, result.get('error'))
            else:
                messages.error(request, 'Your Transaction has failed.')
        except Exception as err:
            return default_undefined_exception(request, err=err, ref=REF, logger=logger)
        return redirect('plans.index')

    @classmethod
    def invoice_pay_with_benefit(cls, request: HttpRequest) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            invoice_id = Utility.decrypt(request.POST.get('invoice_id'))
            invoice = Invoice.objects.get(id=invoice_id)
            user = User.objects.get(id=invoice.created_by)
            settings_dict = Utility.settings_by_id(user.id)
            company_settings = Utility.get_company_payment_setting(user.id)
            secret_key = company_settings.get('benefit_secret_key')
            amount = float(request.POST.get('amount'))
            if amount > invoice.get_due():
                messages.error(request, 'Invalid amount.')
                return redirect(get_redirect_url(request))
            payload = {
                'amount': amount,
                'currency': settings_dict.get('site_currency', 'BHD'),
                'customer_initiated': True,
                'threeDSecure': True,
                'save_card': False,
                'description': invoice.invoice_id,
                'metadata': {'udf1': 'Metadata 1'},
                'reference': {'transaction': 'txn_01', 'order': 'ord_01'},
                'receipt': {'email': True, 'sms': True},
                'customer': {
                    'first_name': user.name,
                    'middle_name': '',
                    'last_name': '',
                    'email': user.email,
                    'phone': {'country_code': 965, 'number': '51234567'}
                },
                'source': {'id': 'src_bh.benefit'},
                'post': {'url': 'https://webhook.site/fd8b0712-d70a-4280-8d6f-9f14407b3bbd'},
                'redirect': {
                    'url': reverse('BenefitPaymentController.get_invoice_payment_status', kwargs={
                        'invoice_id': invoice_id,
                        'amount': amount
                    })
                }
            }
            res = requests.post(
                'https://api.tap.company/v2/charges',
                headers={
                    'Authorization': f'Bearer {secret_key}',
                    'accept': 'application/json',
                    'content-type': 'application/json',
                },
                data=json.dumps(payload)
            )
            return redirect(res.json()['transaction']['url'])
        except Exception as err:
            return default_undefined_exception(request, err=err, ref=REF, logger=logger)

    @classmethod
    def get_invoice_payment_status(cls, request: HttpRequest,
                                   invoice_id: int, amount: float) -> HttpResponse:
        CNAME = cls.__name__
        MNAME = inspect.currentframe().f_code.co_name
        REF = f"{CNAME}::{MNAME}"
        try:
            invoice = Invoice.objects.get(id=invoice_id)
            order_id = Utility.generate_order_id()
            user = User.objects.get(id=invoice.created_by)
            company_settings = Utility.get_company_payment_setting(user.id)
            secret_key = company_settings.get('benefit_secret_key')
            settings_dict = Utility.settings_by_id(user.id)
            tap_id = request.GET.get('tap_id')
            res = requests.get(
                f'https://api.tap.company/v2/charges/{tap_id}',
                headers={
                    'Authorization': f'Bearer {secret_key}',
                    'accept': 'application/json'
                }
            )
            data = res.json()
            status_code = data['gateway']['response']['code']
            if status_code == '00':
                InvoicePayment.objects.create(
                    invoice_id=invoice_id,
                    date=Utility.today_str(),
                    amount=amount,
                    account_id=0,
                    payment_method=0,
                    order_id=order_id,
                    payment_type='Benefit',
                    receipt='',
                    reference='',
                    description=f'Invoice {Utility.invoice_number_format(settings_dict, invoice.invoice_id)}'
                )
                due = invoice.get_due() - amount
                invoice.status = 3 if due == 0 else 2
                invoice.save()
                Utility.update_user_balance('customer', invoice.customer_id, amount, 'debit')
                customer = Customer.objects.get(id=invoice.customer_id)
                notification = {
                    'payment_price': amount,
                    'invoice_payment_type': 'Benefit',
                    'customer_name': customer.name
                }
                if settings_dict.get('payment_notification'):
                    Utility.send_slack_msg('new_invoice_payment', notification, invoice.created_by)
                if settings_dict.get('telegram_payment_notification'):
                    Utility.send_telegram_msg('new_invoice_payment', notification, invoice.created_by)
                if settings_dict.get('twilio_payment_notification'):
                    Utility.send_twilio_msg(
                        customer.contact,
                        'new_invoice_payment',
                        notification,
                        invoice.created_by
                    )
                webhook = Utility.webhook_setting('New Invoice Payment', invoice.created_by)
                if webhook:
                    status = Utility.webhook_call(
                        webhook['url'],
                        json.dumps(notification),
                        webhook['method']
                    )
                    if status:
                        messages.error(request, 'Transaction has been failed.')
                        return redirect('InvoiceController.copy', invoice_id)
                    else:
                        messages.error(request, 'Webhook call failed.')
                        return redirect(get_redirect_url(request))
                messages.success(request, 'Invoice paid Successfully!')
                return redirect('InvoiceController.copy', invoice_id)
            messages.error(request, 'Transaction fail!')
            return redirect('InvoiceController.copy', invoice_id)
        except Exception as err:
            return default_undefined_exception(request, err=err, ref=REF, logger=logger)
